### PR TITLE
GC Docker containers and images after the container exits

### LINKED ISSUE #973: docker containers are not removed when status is dead
### Nomad version
v0.3.1

### Operating system and Environment details
centos 7 3.10.0-327.10.1.el7.x86_64
Docker version 1.10.3, build 20f81dd

### Issue
When I call nomad stop on a *batch* job, containers associated with dead allocations are not cleaned up. However, if the allocation is still running, the container is cleaned up.

Is it possible Kill is only called on running allocations and thus cleanup only happens for running containers? 

### Reproduction steps
1. Create a container that sleeps for 30 seconds (or long enough you can call nomad stop before it finishes)
2. Create job1 that runs this container as a *batch* type job. 
    - Run job1 and let it complete, status should be dead. 
    - take note of the node the container ran on
    - run nomad stop job1
    - inspect docker on the node that the container ran on, ssh into the node and run `docker ps -a`, the container will be there with a status of exited
3. Create job2 based on job1
    - Run job2 long enough for the container to start executing.
    - take note of the node the container is running on
    - before the job completes, call nomad stop job2
    - inspect docker on the node that the container ran on, ssh into the node and run `docker ps -a`, there won't be any containers, in this case the container is cleaned up.

### PR DESCRIPTION
**Pull request by diptanu**
Fixes #973

### CODE REVIEW

**Comment by sheerun** [comment]
How does it relate to following? Are logs still kept?

> The idea behind keeping the allocations after they have completed for some time is to allow users to inspect the allocation once they have finished in case they want to debug something.

**Comment by dadgar** [comment]
@sheerun They are unrelated. This is about cleaning up the container which is managed by docker while logs are managed by us.

**Comment by diptanu** [comment]
@sheerun Yes, like @dadgar said logs would still be retained as they live in the allocation directory and we don't ask Docker to remove the volumes when we remove the container.

File: client/driver/docker_test.go [hunk 1]
```diff
@@ -757,3 +757,52 @@ func TestDockerUser(t *testing.T) {
...
+	case res := <-handle.WaitCh():
+		if !res.Successful() {
+			t.Fatalf("err: %v", res)
+		}
+
+		time.Sleep(3 * time.Second)
+
+		// Ensure that the container isn't present
+		info, err := client.InspectContainer(handle.(*DockerHandle).containerID)
+		fmt.Printf("INFO %#v \n", info)
```
Comment (dadgar) [inline client/driver/docker_test.go]: Remove debug print

File: client/driver/docker.go [hunk 2]
```diff
@@ -893,4 +852,26 @@ func (h *DockerHandle) run() {
...
+	if err := h.client.StopContainer(h.containerID, 0); err != nil {
+		_, noSuchContainer := err.(*docker.NoSuchContainer)
+		_, containerNotRunning := err.(*docker.ContainerNotRunning)
+		if !containerNotRunning && !noSuchContainer {
+			h.logger.Printf("[ERR] driver.docker: error stopping container: %v", err)
+		}
+	}
+
+	// Remove the container
+	if err := h.client.RemoveContainer(docker.RemoveContainerOptions{ID: h.containerID, Force: true}); err != nil {
```
Comment (dadgar) [inline client/driver/docker.go]: Wrap this in `if h.cleanupContainer {`

The user can configure Nomad to keep the container

File: client/driver/docker.go [hunk 3]
```diff
@@ -893,4 +852,26 @@ func (h *DockerHandle) run() {
...
+		h.logger.Printf("[ERR] driver.docker: error removing container: %v", err)
+	}
+
+	// Cleanup the image
+	if h.cleanupImage {
+		if err := h.client.RemoveImage(h.imageID); err != nil {
```
Comment (dadgar) [inline client/driver/docker.go]: Why did you get rid of the logic that was here before?

File: client/driver/docker.go (hunk 2 above)
Comment (diptanu) [inline client/driver/docker.go]: @dadgar There is no point in keeping a container which has exits. The only reason I can think about keeping a container once it exits is to see the logs, which we copy in alloc dirs now.

File: client/driver/docker.go (hunk 3 above)
Comment (diptanu) [inline client/driver/docker.go]: @dadgar The extra logic's purpose was to only tell the user why an image couldn't be removed. The error from `RemoveImage` API conveys that message anyways and we are not returning this error anyways.

File: client/driver/docker.go (hunk 2 above)
Comment (kurtwheeler) [reply client/driver/docker.go]: It looks like you didn't do this... Is there any way to get Nomad to keep a container that has exited? There is a reason to keep a container once it has exited besides logs: I want to run `docker inspect` on it so I can tell why it was killed. All nomad is telling me is `Exit Code: 137, Exit Message: "Docker container exited with non-zero exit code: 137"` which isn't OOMKilled, (that shows up in the exit message). The logs don't have any hints, so I'm trying to debug further but all the containers disappear as soon as they're mysteriously killed.

**Pull request merged**